### ISSUE TITLE
How to overwrite variables?

### ISSUE THREAD

**Issue by AdrienWalkowiak** [open-policy-agent/opa#1500]
Hi,

I am trying to return a custom message when my rule is broken, depending on which test failed, for an improved user experience. I did not find a way to overwrite my message variable so far, any idea how this could be achieve in opa?

For instance:
# Expected Behavior

myrule[{"msg": message}] {
    message := "some default message"
    sometestfunction
    message := "new value for message since sometestfunction succeeded"
}
sometestfunction(){ 1 != 1 }

>> in this case message should be set to "some default message" since sometestfunction failed
and with this:

myrule[{"msg": message}] {
    message := "some default message"
    sometestfunction
    message := "new value for message since sometestfunction succeeded"
}
sometestfunction(){ 1 == 1 }

>> this would have message be overwritten

# Actual Behavior
in both cases I get: "1 error occurred: 4:5: rego_compile_error: var message assigned above"

# Additional Info
I tried to play with the "=" operator too, but it does not seem to help, since the variable assignment is done only once with this too.

If there is no syntax to overwrite a variable in rego, is there a workaround for my specific use case?

Thank you,

Adrien

**Comment by patrick-east** [comment]
One important thing to consider is that with Rego those rules are not being processed in an imperative way. When the rule is being evaluated with two `message := ...` statements you are saying to bind `message` to >1 thing for the result, hence the error. If you use `message = ...` with two string constants it will fail with a similar problem because it is either assignment or equality.

The other issue is that as soon as `sometestfunction` fails the rule body evaluates to false and `myrule` isn't going to be updated with the new `{"msg": "..."}` object. This is probably the biggest problem with the current examples since it wouldn't really matter so much what the message was.

With that in mind you may want to consider formatting this a little bit differently. To keep the per-test messaging similar to the examples you could look into using the `else` keyword https://www.openpolicyagent.org/docs/latest/how-do-i-write-policies/#else-keyword. Something like https://play.openpolicyagent.org/p/sdYbC2dVcE shows using it with your example. When I was putting that together I found that the syntax for else didn't like having the `myrule[{"msg": message}]` head so I split it into a different rule just for that one message. There might be a better way to write that (my rego-foo is still low level). But the gist is that with that you can, when the test function is false, assign another value instead in the else block.

Another option is that you could do something like https://play.openpolicyagent.org/p/iXc0KmbGE2 with the `default` keyword. This is going to assign it to the message specified in the test rule if `sometestfunction` is true, or it will get the `default` message. You could flip it around and have a message for when the test fails instead like: https://play.openpolicyagent.org/p/b1LG3JtIlE

I'm assuming that there are other solutions for this too, @tsandall and others will likely have some better ideas.

**Comment by AdrienWalkowiak** [comment]
Thank you very much this makes more sense now.

I think my main misconception was that rules **and** functions where being processed in an imperative way. I will try to use you else example as it seems to be doing exactly what I needed :)

Thank you!
